tuplex/codegen/src/LLVMEnvironment.cc

```diff
@@ -357,6 +361,131 @@ namespace tuplex {
+            using namespace llvm;
+
+            if(iterableType == python::Type::EMPTYLIST || iterableType == python::Type::EMPTYTUPLE) {
+                // use dummy type for empty iterator
+                return i64Type();
+            }
+
+            auto it = _generatedIterIteratorTypes.find(iterableType);
+            if(_generatedIterIteratorTypes.end() != it) {
+                return it->second;
+            }
+
+            std::string iteratorName;
+            std::vector<llvm::Type*> memberTypes;
+            // iter iterator struct: { pointer to block address (i8*), current index (i64 for range otherwise i32), pointer to iterable struct type,
+            // iterable length (for string and tuple)}
+            memberTypes.push_back(llvm::Type::getInt8PtrTy(_context, 0));
+            if(iterableType == python::Type::RANGE) {
+                iteratorName = "range_";
+                memberTypes.push_back(llvm::Type::getInt64Ty(_context));
+                memberTypes.push_back(llvm::PointerType::get(getRangeObjectType(), 0));
+            } else {
+                memberTypes.push_back(llvm::Type::getInt32Ty(_context));
+                if(iterableType.isListType()) {
+                    iteratorName = "list_";
+                    memberTypes.push_back(llvm::PointerType::get(getListType(iterableType), 0));
+                } else if(iterableType == python::Type::STRING) {
+                    iteratorName = "str_";
+                    memberTypes.push_back(llvm::Type::getInt8PtrTy(_context, 0));
+                    memberTypes.push_back(llvm::Type::getInt64Ty(_context));
+                } else if(iterableType.isTupleType()) {
+                    iteratorName = "tuple_";
+                    memberTypes.push_back(llvm::PointerType::get(getOrCreateTupleType(flattenedType(iterableType)), 0));
+                    memberTypes.push_back(llvm::Type::getInt64Ty(_context));
+                }
+            }
+
+            llvm::ArrayRef<llvm::Type*> members(memberTypes);
+            auto iteratorContextType = llvm::StructType::create(_context, members, "struct." + iteratorName + twine);
+            _generatedIterIteratorTypes[iterableType] = iteratorContextType;
+            return iteratorContextType;
+        }
+
+        llvm::Type *LLVMEnvironment::getZipIteratorType(const python::Type &argsType, const std::vector<IteratorInfo *> &argsIteratorInfo, const std::string &twine) {
+            using namespace llvm;
+
+            if(argsType.parameters().empty()) {
+                return i64Type();
+            }
+
+            std::vector<llvm::Type*> memberTypes;
+            // zip iterator struct: { pointer to each child iterator struct type }
+            for (int i = 0; i < argsType.parameters().size(); ++i) {
+                auto iterableType = argsType.parameters()[i];
+                if(iterableType == python::Type::EMPTYITERATOR || iterableType == python::Type::EMPTYLIST || iterableType == python::Type::EMPTYTUPLE) {
```

Comment:
add comment to explain logic here, i.e. why i64Type is returned.

Reply:
If any argument is empty, zip returns an empty iterator. i64Type is used for empty iterator.
Added comments.